sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,617 @@
+    cgh.single_task([=](sycl::kernel_handler kh) {
+      int i = kh.get_specialization_constant<spec_var>();
+    });
+  });
+}
+```
+
+Device global variables, by contrast, are referenced by their address:
+
+```
+sycl::ext::oneapi::device_global<int> dev_var;
+
+void func(sycl::queue q) {
+  int val = 42;
+  q.copy(&val, dev_var).wait();      // The 'dev_var' parameter is by reference
+  q.submit([&](sycl::handler &cgh) {
+    cgh.single_task([=] {
+      int i = dev_var;
+    });
+  });
+}
+```
+
+This is a key difference because the compiler does not statically know which
+device global variable is being referenced; we only know the address at
+runtime.  As we will see later, this has a ramification on the integration
+headers and on the mechanism that connects instances of device global variables
+in host code with their corresponding instances in device code.
+
+Another issue relates to the `device_image_scope` property which can be applied
+to a device global variable declaration.  The intent of this property is to
+allow a device global variable to be implemented directly on top of a SPIR-V
+module scope global variable.  When this property is **not** present, an
+instance of a device global variable is shared across all device images that
+are loaded onto a particular device.  By contrast, when this property **is**
+present, each device image has its own instance of the device global variable.
+However, since multiple variable instances have confusing semantics, the API
+requires the user to ensure that each such variable exists in exactly one
+device image.  The extension specification has more details on this property.
+
+The important impact on the design, though, is that device global variables
+declared with the `device_image_scope` property have an implementation that is
+quite different from device global variables that are not declared with this
+property.  The sections below describe both implementations.
+
+
+## Design
+
+### Changes to DPC++ headers
+
+#### Class specializations based on `device_image_scope`
+
+The headers, of course, include the declaration of the new `device_global`
+class, which is described in the [extension specification][1].  The declaration
+of this class uses partial specialization to define the class differently
+depending on whether it has the `device_image_scope` property.  When the
+property is not present, the class has a member variable which is a pointer to
+the underlying type.  Member functions which return a reference to the value
+(e.g. `get`) return the value of this pointer:
+
+```
+template<typename T, typename PropertyListT>
+class device_global {
+  T *usmptr;
+ public:
+  T& get() noexcept { return *usmptr; }
+  /* other member functions */
+};
+```
+
+However, when the property is present, it has a member variable which is the
+type itself, and member functions return a reference to this value.
+
+```
+template<typename T, typename PropertyListT>
+class device_global {
+  T val;
```

Comment:
```suggestion
  T val{};
```
would avoid having a default constructor?

Reply:
And what if we do not want default initialization because it is an array of 128GiB? I guess we can just provide a `T` where there is no initialization, like https://github.com/keryell/teaching_C-plus-plus/blob/master/non-initializing/remove_initialization.cpp

Reply:
We discussed offline and are going to argue not to change anything now regarding .  The union approach that you show can prevent default constructors of the type from running, but zero init still happens on the union if it has static storage duration.  So constructors can be prevented with the union, but zero init will still occur which may have cost depending on size and the device.

We currently think that zero init of `device_global` is worth the rare cost, to mirror C++ semantics and to treat the object as global-like.  We propose to add a `no_init` property or `init_mode` enum value if use cases arise where this matters in practice.  Providing a control that breaks C++ global init semantics is making people nervous, hence the thought that we should wait until a need is proven :)

Reply:
> would avoid having a default constructor?

Yes, that makes sense.  Changed in 59852035c566e4e9b75623aece866fd98b09c534